components/brave_extension/extension/brave_extension/content.ts

```diff
@@ -1,10 +1,100 @@
+// Notify the background script as soon as the content script has loaded.
+// chrome.tabs.insertCSS may sometimes fail to inject CSS in a newly navigated
+// page when using the chrome.webNavigation API.
+// See: https://bugs.chromium.org/p/chromium/issues/detail?id=331654#c15
+// The RenderView should always be ready when the content script begins, so
+// this message is used to trigger CSS insertion instead.
+chrome.runtime.sendMessage({
+  type: 'pageContentReadyForInjection'
+})
+
 const unique = require('unique-selector').default
+
 let target: EventTarget | null
+let genericExceptions: Array<string> | undefined = undefined
+let classIdBuffer: { classes: string[], ids: string[] } = { classes: [], ids: [] }
+
+const queriedIds = new Set()
+const queriedClasses = new Set()
+const regexWhitespace = /\s/
 
 function getCurrentURL () {
   return window.location.hostname
 }
 
+const getClassesAndIds = function (addedNodes: Element[]) {
+  const ids = []
+  const classes = []
+
+  for (const node of addedNodes) {
+    let nodeId = node.id
+    if (nodeId && nodeId.length !== 0) {
+      nodeId = nodeId.trim()
+      if (!queriedIds.has(nodeId) && nodeId.length !== 0) {
+        ids.push(nodeId)
+        queriedIds.add(nodeId)
+      }
+    }
+    let nodeClass = node.className
+    if (nodeClass && nodeClass.length !== 0 && !regexWhitespace.test(nodeClass)) {
+      if (!queriedClasses.has(nodeClass)) {
+        classes.push(nodeClass)
+        queriedClasses.add(nodeClass)
+      }
+    } else {
+      let nodeClasses = node.classList
+      if (nodeClasses) {
+        let j = nodeClasses.length
+        while (j--) {
+          const nodeClassJ = nodeClasses[j]
+          if (queriedClasses.has(nodeClassJ) === false) {
+            classes.push(nodeClassJ)
+            queriedClasses.add(nodeClassJ)
+          }
+        }
+      }
+    }
+  }
+  return { classes, ids }
+}
+
+const handleNewNodes = (newNodes: Element[]) => {
+  const { classes, ids } = getClassesAndIds(newNodes)
+  if (!(genericExceptions)) {
+    classIdBuffer.classes.push(...classes)
+    classIdBuffer.ids.push(...ids)
+  } else {
+    chrome.runtime.sendMessage({
```

Comment:
In practice, do you know how often `classIdStylesheet` gets hit for pages?  Just making sure we're not inserting like 1000 stylesheets.  I think maybe we should be debouncing these calls.

Reply:
It depends entirely on how often the site makes mutations to the page. I can try on a few different ones if you'd like specific examples.
Note that this is already batched on a per-mutation basis - i.e. if a node tree is added, it's only called once for all the classes and ids in the entire tree.